File: php/src/Domain/Rules/Entity/CompatibilityRule.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Rules\Entity;

use Alxarafe\App\Domain\Rules\ValueObject\AttributeId;
use Alxarafe\App\Domain\Rules\ValueObject\CompatibilityRuleId;
use Alxarafe\App\Domain\Rules\ValueObject\RuleScope;
use Alxarafe\App\Domain\Rules\ValueObject\RuleType;
use InvalidArgumentException;

/**
 * Entity representing a warehouse compatibility or storage constraint rule.
 *
 * Rules configure compatibility, e.g., a COLD merchandise attribute
 * REQUIRES a REFRIGERATED location attribute at the LOCATION scope, or
 * a HAZMAT merchandise attribute FORBIDS a standard storage location.
 */
final readonly class CompatibilityRule
{
    public function __construct(
        private CompatibilityRuleId $id,
        private RuleType $ruleType,
        private AttributeId $sourceAttributeId, // Usually an ItemFamily attribute
        private AttributeId $targetAttributeId, // Usually a Location attribute
        private RuleScope $scope,
    ) {
        if ($sourceAttributeId->equals($targetAttributeId)) {
            throw new InvalidArgumentException(
                'Source and target attributes cannot be the same in a CompatibilityRule.'
            );
        }
    }

    public function id(): CompatibilityRuleId
    {
        return $this->id;
    }

    public function ruleType(): RuleType
    {
        return $this->ruleType;
    }

    public function sourceAttributeId(): AttributeId
    {
        return $this->sourceAttributeId;
    }

    public function targetAttributeId(): AttributeId
    {
        return $this->targetAttributeId;
    }

    public function scope(): RuleScope
    {
        return $this->scope;
    }
}

File: java/src/main/java/com/alxarafe/app/domain/rules/entity/CompatibilityRule.java
package com.alxarafe.app.domain.rules.entity;

import com.alxarafe.app.domain.rules.valueobject.AttributeId;
import com.alxarafe.app.domain.rules.valueobject.CompatibilityRuleId;
import com.alxarafe.app.domain.rules.valueobject.RuleScope;
import com.alxarafe.app.domain.rules.valueobject.RuleType;

import java.util.Objects;

/**
 * Entity representing a warehouse compatibility or storage constraint rule.
 *
 * Rules configure compatibility, e.g., a COLD merchandise attribute
 * REQUIRES a CHILLED location attribute at the LOCATION scope, or
 * a HAZMAT merchandise attribute FORBIDS a standard storage location.
 */
public final class CompatibilityRule {

    private final CompatibilityRuleId id;
    private final RuleType ruleType;
    private final AttributeId sourceAttributeId;
    private final AttributeId targetAttributeId;
    private final RuleScope scope;

    public CompatibilityRule(CompatibilityRuleId id, RuleType ruleType,
                             AttributeId sourceAttributeId, AttributeId targetAttributeId,
                             RuleScope scope) {
        this.id = Objects.requireNonNull(id, "CompatibilityRule id cannot be null.");
        this.ruleType = Objects.requireNonNull(ruleType, "CompatibilityRule ruleType cannot be null.");
        this.sourceAttributeId = Objects.requireNonNull(sourceAttributeId, "CompatibilityRule sourceAttributeId cannot be null.");
        this.targetAttributeId = Objects.requireNonNull(targetAttributeId, "CompatibilityRule targetAttributeId cannot be null.");
        this.scope = Objects.requireNonNull(scope, "CompatibilityRule scope cannot be null.");
        if (sourceAttributeId.equals(targetAttributeId)) {
            throw new IllegalArgumentException(
                    "Source and target attributes cannot be the same in a CompatibilityRule.");
        }
    }

    public CompatibilityRuleId id() {
        return id;
    }

    public RuleType ruleType() {
        return ruleType;
    }

    public AttributeId sourceAttributeId() {
        return sourceAttributeId;
    }

    public AttributeId targetAttributeId() {
        return targetAttributeId;
    }

    public RuleScope scope() {
        return scope;
    }
}
